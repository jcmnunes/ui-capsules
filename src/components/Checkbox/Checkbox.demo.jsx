import React, { useState } from 'react';
import Select from '../../lib/Select/Select';
import Checkbox from '../../lib/Checkbox/Checkbox';
import Input from '../../lib/Input/Input';
import Demo, { Booleans, StyledCheckbox, StyledSelect } from '../common/Demo';
import { GITHUB_URL, SIZE_OPTS, APPEARANCE_OPTS } from './Checkbox.constants';

const CheckboxDemo = () => {
  const [checked, setChecked] = useState(false);
  const [size, setSize] = useState(SIZE_OPTS[0]);
  const [children, setChildren] = useState('Hello world');
  const [appearance, setAppearance] = useState(APPEARANCE_OPTS[2]);

  const Component = (
    <Checkbox checked={checked} size={size.value} appearance={appearance.value}>
      {children}
    </Checkbox>
  );

  return (
    <Demo codeURL={GITHUB_URL} component={Component}>
      <>
        <Booleans>
          <StyledCheckbox
            checked={checked}
            onChange={() => setChecked(!checked)}
            appearance="success"
          >
            <pre>checked</pre>
          </StyledCheckbox>
        </Booleans>
        <StyledSelect>
          <pre>children</pre>
          <Input
            isBlock
            type="text"
            value={children}
            onChange={ev => setChildren(ev.target.value)}
          />
        </StyledSelect>
        <StyledSelect>
          <pre>size</pre>
          <Select defaultValue={SIZE_OPTS[0]} options={SIZE_OPTS} onChange={opt => setSize(opt)} />
        </StyledSelect>
        <StyledSelect>
          <pre>appearance</pre>
          <Select
            defaultValue={APPEARANCE_OPTS[2]}
            options={APPEARANCE_OPTS}
            onChange={opt => setAppearance(opt)}
          />
        </StyledSelect>
      </>
    </Demo>
  );
};

export default CheckboxDemo;
