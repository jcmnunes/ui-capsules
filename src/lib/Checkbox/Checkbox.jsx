import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/macro';

const getColors = props => {
  const colors = {};
  switch (props.intent) {
    case 'primary':
      colors.background = props.theme.blue500;
      break;
    case 'success':
      colors.background = props.theme.green500;
      break;
    case 'neutral':
      colors.background = props.theme.neutral300;
      break;
    default:
      colors.background = props.theme.neutral300;
      break;
  }
  return colors;
};

const CustomCheckbox = styled.span`
  position: absolute;
  border-radius: 4px;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: ${props => props.theme.neutral100};
  transition: background-color 0.2s ease;

  &:after {
    content: '';
    position: absolute;
    display: none;
    left: 9px;
    top: 5px;
    width: 7px;
    height: 12px;
    border: solid white;
    border-width: 0 3px 3px 0;
    transform: rotate(45deg);
  }

  &:hover {
    background-color: ${props => props.theme.neutral200};
  }
`;

const Label = styled.label`
  display: inline-flex;
  height: 25px;
  align-items: center;
  position: relative;
  padding-left: 35px;
  cursor: pointer;
  font-size: 18px;
  user-select: none;
`;

const StyledInput = styled.input`
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;

  &:focus {
    & ~ ${CustomCheckbox} {
      outline: none;
      box-shadow: 0 0 0 4px ${props => props.theme.blue100};
    }
  }

  &:checked {
    & ~ ${CustomCheckbox} {
      background-color: ${props => getColors(props).background};

      &:after {
        display: block;
      }
    }
  }
`;

const Checkbox = ({ checked, onChange, children, intent, onClick, ...other }) => (
  <Label onClick={onClick} {...other}>
    {children}
    <StyledInput type="checkbox" checked={checked} onChange={onChange} intent={intent} />
    <CustomCheckbox />
  </Label>
);

Checkbox.defaultProps = {
  children: '',
  intent: 'neutral',
  onClick: null,
};

Checkbox.propTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  children: PropTypes.node,
  intent: PropTypes.oneOf(['primary', 'success', 'neutral']),
  onClick: PropTypes.func,
};

export default Checkbox;
