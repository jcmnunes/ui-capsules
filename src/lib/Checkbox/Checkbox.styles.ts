import { styled } from '../stitches.config';

export const Label = styled('label', {
  display: 'flex',
  alignItems: 'center',
  position: 'relative',
  userSelect: 'none',
  cursor: 'pointer',
  fontWeight: 500,
  color: '$neutral700',

  variants: {
    disabled: {
      true: {
        color: '$neutral400',
        cursor: 'initial',
      },
    },

    size: {
      medium: {
        height: 18,
        fontSize: '$2',
        paddingLeft: 24,
        lineHeight: '18px',
      },

      large: {
        height: 25,
        fontSize: '$2',
        paddingLeft: 35,
        lineHeight: '25px',
      },
    },
  },
});

export const CustomCheckbox = styled('span', {
  position: 'absolute',
  borderRadius: '$medium',
  top: 0,
  left: 0,
  backgroundColor: '$bg',
  borderWidth: 1,
  border: 'none',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  boxShadow: '$border',

  '&:hover': {
    backgroundColor: '$neutral50',
  },

  '&:after': {
    content: '""',
    position: 'absolute',
    display: 'none',
    border: 'solid white',
    transform: 'rotate(45deg)',
  },

  variants: {
    disabled: {
      true: {
        background: '$neutral300',
        boxShadow: 'none',
        pointerEvents: 'none',
      },
    },

    size: {
      medium: {
        width: 18,
        height: 18,

        '&:after': {
          left: '6px',
          top: '3px',
          width: 6,
          height: 10,
          borderWidth: '0 2px 2px 0',
        },
      },

      large: {
        width: 24,
        height: 24,

        '&:after': {
          top: 4,
          left: 8,
          width: 8,
          height: 13,
          borderWidth: '0 3px 3px 0',
        },
      },
    },
  },
});

export const StyledInput = styled('input', {
  position: 'absolute',
  opacity: 0,
  cursor: 'pointer',
  height: 0,
  width: 0,

  '&[data-focus-visible-added]:focus': {
    '& ~ .uic-custom-checkbox': {
      outline: 'none',
      boxShadow: '$focus',
    },
  },

  '&:checked': {
    '& ~ .uic-custom-checkbox': {
      backgroundColor: '$primary500',
      borderColor: '$primary500',
      boxShadow: '$borderPrimary',

      '&:after': {
        display: 'block',
      },
    },

    '&:disabled': {
      '& ~ .uic-custom-checkbox': {
        opacity: 0.7,
      },
    },
  },
});
