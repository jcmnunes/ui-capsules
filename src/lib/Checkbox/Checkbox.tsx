import React, { ComponentPropsWithRef, forwardRef } from 'react';
import { Box } from '../Box/Box';
import { CSSProp, DualSize } from '../types';
import { CustomCheckbox, Label, StyledInput } from './Checkbox.styles';

interface Props extends Omit<ComponentPropsWithRef<'input'>, 'size'>, CSSProp {
  size?: DualSize;
}

export const Checkbox = forwardRef<HTMLInputElement, Props>(
  ({ checked, size = 'medium', disabled = false, children, onChange, css, ...rest }, ref) => (
    <Box css={css}>
      <Label size={size} onClick={e => e.stopPropagation()} disabled={disabled}>
        {children && <span>{children}</span>}

        <StyledInput
          ref={ref}
          type="checkbox"
          checked={checked}
          onChange={onChange}
          disabled={disabled}
          {...rest}
        />

        <CustomCheckbox className="uic-custom-checkbox" size={size} disabled={disabled} />
      </Label>
    </Box>
  ),
);

Checkbox.displayName = 'Checkbox';
Checkbox.defaultProps = {
  size: 'medium',
  disabled: false,
};
