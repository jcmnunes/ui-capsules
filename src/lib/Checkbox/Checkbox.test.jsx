import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Checkbox } from './Checkbox';

describe('Checkbox', () => {
  it('should have a displayName', () => {
    expect(Checkbox.displayName).toBe('Checkbox');
  });

  it('should render with no errors', () => {
    const { container } = render(
      <Checkbox checked onChange={() => {}}>
        checkbox label
      </Checkbox>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the checkbox checked when passing checked prop as true', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <Checkbox checked onChange={onChange}>
        checkbox label
      </Checkbox>,
    );
    const checkbox = getByLabelText(/checkbox label/i);
    expect(checkbox.checked).toBe(true);
  });

  it('should have the checkbox checked when passing checked prop as false', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <Checkbox checked={false} onChange={onChange}>
        checkbox label
      </Checkbox>,
    );
    const checkbox = getByLabelText(/checkbox label/i);
    expect(checkbox.checked).toBe(false);
  });

  it('should call the passed onChange function when clicking the label', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <Checkbox checked onChange={onChange}>
        checkbox label
      </Checkbox>,
    );
    const label = getByTestId('label');
    fireEvent.click(label);
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('should be a controlled component', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <Checkbox checked onChange={onChange}>
        checkbox label
      </Checkbox>,
    );
    const checkbox = getByLabelText(/checkbox label/i);
    fireEvent.click(checkbox);
    // Checkbox should not be unchecked because the checked
    // prop is set to true
    expect(checkbox.checked).toBe(true);
  });
});
