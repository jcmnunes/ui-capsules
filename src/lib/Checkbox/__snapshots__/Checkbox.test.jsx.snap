// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with no errors 1`] = `
.c3 {
  position: absolute;
  border-radius: 4px;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: #fff;
  -webkit-transition: background-color 0.2s ease;
  transition: background-color 0.2s ease;
  border: 1px solid #E4E7EB;
}

.c3:after {
  content: '';
  position: absolute;
  display: none;
  left: 5px;
  top: 2px;
  width: 5px;
  height: 9px;
  border: solid white;
  border-width: 0 2px 2px 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3:hover {
  background-color: #F5F7FA;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  padding-left: 22px;
  cursor: pointer;
  font-size: 16px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: inherit;
}

.c1 {
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;
}

.c1:focus ~ .c2 {
  outline: none;
  box-shadow: 0 0 0 2px #B6E0FE;
}

.c1:checked ~ .c2 {
  background-color: #9AA5B1;
  border-color: #9AA5B1;
}

.c1:checked ~ .c2:after {
  display: block;
}

.c1:checked:disabled ~ .c2 {
  background-color: #E4E7EB;
  border-color: #E4E7EB;
}

<label
  class="c0"
  data-testid="label"
>
  checkbox label
  <input
    checked=""
    class="c1"
    type="checkbox"
  />
  <span
    class="c2 c3"
  />
</label>
`;
